pkg/controller/plan/adapter/vsphere/builder.go

```diff
@@ -241,9 +243,34 @@ func (r *Builder) mapMacStaticIps(vm *model.VM) string {
 		return ""
 	}
 	configurations := []string{}
+	gatewaySet := false
 	for _, guestNetwork := range vm.GuestNetworks {
 		if guestNetwork.Origin == string(types.NetIpConfigInfoIpAddressOriginManual) {
-			configurations = append(configurations, fmt.Sprintf("%s:ip:%s", guestNetwork.MAC, guestNetwork.IP))
+			gateway := ""
+			if !gatewaySet {
+				isIpv4 := net.IP.To4(net.ParseIP(guestNetwork.IP)) != nil
+				for _, ipStack := range vm.GuestIpStacks {
+					if net.IP.To4(net.ParseIP(ipStack.Gateway)) != nil && !isIpv4 || net.IP.To4(net.ParseIP(ipStack.Gateway)) == nil && isIpv4 {
+						// not the right IPv4 / IPv6 correlation
+						continue
+					}
+					network, err := netip.ParsePrefix(fmt.Sprintf("%s/%d", guestNetwork.IP, guestNetwork.PrefixLength))
+					if err != nil {
+						return ""
+					}
+					gw, err := netip.ParseAddr(ipStack.Gateway)
+					if err != nil {
+						return ""
+					}
+					if network.Contains(gw) {
+						// checks if the gateway in the right network subnet. we set gateway only once as it is suppose to be system wide.
+						gateway = ipStack.Gateway
+						gatewaySet = true
+					}
+				}
+			}
+			dnsString := strings.Join(guestNetwork.DNS, ",")
+			configurations = append(configurations, fmt.Sprintf("%s:ip:%s,%s,%d,%s", guestNetwork.MAC, guestNetwork.IP, gateway, guestNetwork.PrefixLength, dnsString))
```

Comment:
do we know what happens with dual stack? as we spoke offline, I'm not sure if that's supported in kubevirt but as far as I remember it is supported in vSphere and if I'm right on this, then could it happen that we set the GW for IPv4 and then override it when specifying IPv6 for the same MAC address?

Reply:
From what I've tested the VM will result in 2 gateways, one for IPv4 and one for IPv6. The DNS list will be combined for both (IPv4 and IPv6). Checking into details of the NIC per version it is set 1:1 to the source. Maybe someone in CNI can answer if it's allowed?

Reply:
@phoracek

Reply:
I had offline discussion with @ahadas, it looks alright.